### PR TITLE
Mutation badge

### PR DESCRIPTION
**Pull request by borNfreee**
This PR adds a Mutation Badge feature

How it looks like (for current branch): [Infection MSI](https://infection.github.io)

Image code is: `[Infection MSI](https://infection.github.io)`

- [x] Covered by tests
- [x] Change `mutation-badge` to `master` before merging 
- [x] Doc PR https://github.com/infection/site/pull/25

Fixes https://github.com/infection/infection/issues/168

To understand how it works, read this: http://stryker-mutator.io/blog/2018-02-08/get-your-mutation-score-badge-now.html

We will use Stryker's Dashboard API. Currently, it will work only for Travis, other CI servers will be implemented on demand.

Configuration is quite easy: 

1. Go to https://dashboard.stryker-mutator.io and sign with your github account.
2. Flip the switch next to your repository (activate it)
3. Configure your API key in your project. It is possible to ncrypt this variable using the encrypted environment variables. For example:
`$ travis encrypt STRYKER_DASHBOARD_API_KEY=89b99910-04d8-4xxx-9x91-23d709x828b4 --add`
4. Configure the badge logger in your `infection.json` file

```json
{
   "logs": {
       "badge": {"branch": "master"}
   }
}
```

### CODE REVIEW

File: src/Http/BadgeApiClient.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+class BadgeApiClient
+{
+    const STRYKER_DASHBOARD_API_URL = 'https://dashboard.stryker-mutator.io/api/reports';
+
+    /**
+     * @var OutputInterface
+     */
+    private $output;
+
+    /**
```
Comment (borNfreee) [inline src/Http/BadgeApiClient.php]: reminder to remove it

File: src/Logger/BadgeLogger.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+namespace Infection\Logger;
+
+use Infection\Http\BadgeApiClient;
+use Infection\Mutant\MetricsCalculator;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class BadgeLogger implements MutationTestingResultsLogger
+{
+    const ENV_INFECTION_BADGE_API_KEY = 'INFECTION_BADGE_API_KEY';
+    /**
```
Comment (sidz) [inline src/Logger/BadgeLogger.php]: could you please put a new line here?

File: src/Process/Listener/MutationTestingResultsLoggerSubscriber.php [hunk 3]
```diff
@@ -72,54 +82,63 @@ public function onMutationTestingFinished(MutationTestingFinished $event)
...
         }
     }
 
     private function filterLogTypes(array $logTypes): array
     {
         $allowedFileTypes = [
-            self::TEXT_FILE,
-            self::DEBUG_FILE,
-            self::SUMMARY_FILE,
+            ResultsLoggerTypes::TEXT_FILE,
```
Comment (sidz) [inline src/Process/Listener/MutationTestingResultsLoggerSubscriber.php]: could you please create a new constant variable in the `ResultsLoggerTypes` which will have all of them?

File: src/Process/Listener/MutationTestingResultsLoggerSubscriber.php [hunk 4]
```diff
@@ -72,54 +82,63 @@ public function onMutationTestingFinished(MutationTestingFinished $event)
...
+            if (!in_array($key, $allowedFileTypes, true)) {
                 unset($logTypes[$key]);
             }
         }
 
         return $logTypes;
     }
 
-    private function useLogger(string $logType)
+    private function useLogger(string $logType, $config)
```
Comment (sidz) [inline src/Process/Listener/MutationTestingResultsLoggerSubscriber.php]: is the second parameter a string?

Comment (borNfreee) [reply src/Process/Listener/MutationTestingResultsLoggerSubscriber.php]: no, each logger has its own config: somewhere it's a string, somewhere it's an object

```json
{
    "logs": {
        "text": "infection-log.txt",
        "badge": {
            "branch": "mutation-badge"
        }
    }
}

File: src/Process/Listener/MutationTestingResultsLoggerSubscriber.php [hunk 5]
```diff
@@ -89,15 +89,8 @@ public function onMutationTestingFinished(MutationTestingFinished $event)
...
-        $allowedFileTypes = [
-            ResultsLoggerTypes::TEXT_FILE,
-            ResultsLoggerTypes::DEBUG_FILE,
-            ResultsLoggerTypes::SUMMARY_FILE,
-            ResultsLoggerTypes::BADGE,
-        ];
-
-        foreach ($logTypes as $key => $value) {
-            if (!in_array($key, $allowedFileTypes, true)) {
+        foreach (ResultsLoggerTypes::ALL as $key => $value) {
```
Comment (sidz) [inline src/Process/Listener/MutationTestingResultsLoggerSubscriber.php]: em, seems like `$logTypes` should be here. Isn't it?

Comment (borNfreee) [reply src/Process/Listener/MutationTestingResultsLoggerSubscriber.php]: indeed, thank you

File: src/Http/BadgeApiClient.php [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    public function __construct(OutputInterface $output)
+    {
+        $this->output = $output;
+    }
+
+    public function sendReport(
+        string $apiKey,
+        string $repositorySlug,
+        string $branch,
+        float $mutationScore)
```
Comment (BackEndTea) [inline src/Http/BadgeApiClient.php]: Could you format this like
```php
    public function sendReport(
        string $apiKey,
        string $repositorySlug,
        string $branch,
        float $mutationScore
    ) {
```

**Comment by borNfreee** [comment]
@sidz @BackEndTea, fixed comments

**Comment by BackEndTea** [comment]
I like it, the only thing that bothers me is that it only works on Travis, and there are people who may use other environments. .

**Comment by borNfreee** [comment]
Majority of open source PHP libs uses Travis I guess. 

We can add other CI servers on demand. 

On the other side, if people use Jenkins in some company (like me for example), I don't think they need mutation badge.

So the target audience for this badge is **GitHub + Travis**, other combinations - on demand.

**Comment by simondel** [comment]
@BackEndTea @borNfreee The same is true for Stryker. We think that mainly open source projects will want the badges and we think that most projects use Travis in the open source world.

If you guys happen to add support for different build platforms in the future, let us know! We could copy the implementation for Stryker 👍

**Comment by borNfreee** [comment]
no problem @simondel

**Comment by borNfreee** [comment]
Thanks for the review @sidz and @BackEndTea

**Comment by nicojs** [comment]
Awesome to see a second mutation testing framework implementing the mutation score badge! We'll keep you informed on any changes we're thinking of doing to the dashboard.

🍻

**Pull request merged**